Deprecate the condition attribute and related functionality

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit deprecates the Instruction.condition and c_if() method for removal in 2.0. This functionality has been superseded by the more sophisiticated `IfElseOp` for some time now. Removing the condition property will simplify the Qiskit data model as it is implemented in a kind of ad-hoc manner that adds additional overhead to manually check it in many places.

For the unittest modifications the deprecation warning for the .condtion getter is suppressed for the entire suite because this gets triggered internally by the transpiler and a lot of other places, including from rust code as until it is removed we need to use it to check that piece of the data model.

### Details and comments

Fixes #9556

mtreinish committed:
Deprecate the condition attribute and related functionality

This commit deprecates the Instruction.condition and c_if() method for
removal in 2.0. This functionality has been superseded by the more
sophisiticated `IfElseOp` for some time now. Removing the condition
property will simplify the Qiskit data model as it is implemented in
a kind of ad-hoc manner that adds additional overhead to manually check
it in many places.

For the unittest modifications the deprecation warning for the .condtion
getter is suppressed for the entire suite because this gets triggered
internally by the transpiler and a lot of other places, including from
rust code as until it is removed we need to use it to check that piece
of the data model.

Fixes #9556

mtreinish committed:
Add missing assertWarns

mtreinish committed:
Merge branch 'main' into deprecate-condition_cif

ElePT commented on qiskit/circuit/instruction.py:
```diff
@@ -160,6 +160,7 @@ def to_mutable(self):
         return self.copy()
 
     @property
+    @deprecate_func(since="1.3.0", removal_timeline="in 2.0.0", is_property=True, stacklevel=2)
```
I believe that 2 is the default stack level, right? It doesn't hurt but it made me think for a second, so maybe:
```suggestion
    @deprecate_func(since="1.3.0", removal_timeline="in 2.0.0", is_property=True)
```

mtreinish replied:
Oh good call, this was leftover from me playing around with the deprecation warning in different settings. We end up accessing this in a few different ways internally and I wanted to see what was user visible or not depending on circuit usage. I tried changing this value for different scenarios and settled on 2 being the best choice and left it explicit like this. I'll remove the kwarg so people realize it's the default.

ElePT commented on releasenotes/notes/deprecate-condition_c_if-9548e5522814fe9c.yaml:
```diff
@@ -0,0 +1,38 @@
+    in Qiskit 2.0 and along with any other uses of this in the Qiskit data
+    model. This functionality has been superseded by the :class:`.IfElseOp`
+    which can be used to describe a classically condition in a circuit. For
+    example, a circuit using :attr:`.Instruction.condition` like::
```
```suggestion
    Deprecated the :attr:`.Instruction.condition` attribute and the
    :meth:`.Instruction.c_if` method. They will be removed
    in Qiskit 2.0, along with any uses in the Qiskit data
    model. This functionality has been superseded by the :class:`.IfElseOp` class
    which can be used to describe a classical condition in a circuit. For
    example, a circuit using  :meth:`.Instruction.c_if` like::
```

ElePT commented on releasenotes/notes/deprecate-condition_c_if-9548e5522814fe9c.yaml:
```diff
@@ -0,0 +1,38 @@
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+
+    can be rewritten using as::
```
```suggestion
    can be rewritten as::
```

mtreinish committed:
Handle deprecation warnings in visual tests too

ElePT reviewed (commented):
Wow, this PR is long. I didn't get to review all tests but I left a few minor suggestions that I saw in the source code/reno.

mtreinish committed:
Apply suggestions from code review

Co-authored-by: Elena Peña Tapia <57907331+ElePT@users.noreply.github.com>

ElePT commented on test/python/transpiler/test_optimize_1q_decomposition.py:
```diff
@@ -212,16 +212,18 @@ def test_ignores_conditional_rotations(self, basis):
         result = passmanager.run(circuit)
-
-        self.assertTrue(Operator(circuit).equiv(Operator(result)))
+        with self.assertWarns(DeprecationWarning):
+            self.assertTrue(Operator(circuit).equiv(Operator(result)))
```
Where is the warning triggered here?

mtreinish replied:
`Operator(circuit)` runs `circuit.to_instruction()` which checks for `condition` iirc.

ElePT commented on test/python/transpiler/test_reset_after_measure_simplification.py:
```diff
@@ -26,12 +26,13 @@ def test_simple(self):
         qc.reset(0)
-
-        new_qc = ResetAfterMeasureSimplification()(qc)
+        with self.assertWarns(DeprecationWarning):
+            new_qc = ResetAfterMeasureSimplification()(qc)
```
Do you think this class raising a deprecation warning is a desired outcome (I'm not sure why it's raised)? Maybe the stack level of the warning isn't correct?

mtreinish replied:
The testing harness will trigger if the warning is raised at any stack level even if it's not user code. This is to let us catch internal usage of deprecated functionality in our dependencies. In this case the `ResetAfterMeasureSimplification` pass only works with `c_if` and `condition`. so the warning is coming from inside the pass.

This does raise a good question about what we want to do with that pass, we probably should convert it to an if statement internally. I don't remember what I was thinking when I opened this 2 weeks ago, probably that I would do it in a followup PR. But it would be better if I do it here instead.

mtreinish replied:
When I went to do this I remember why I didn't change it. I decided at the time that leaving the pass using `c_if` for 1.x made the most sense as using an `IfElseOp` instead could have been viewed as a potential change in behavior and wanted to save that for 2.0.

ElePT replied:
I agree with not changing the behavior yet. I brought this up thinking about downstream packages and how they sometimes get "flooded" by our deprecation warnings even if they aren't actionable. But I guess it's on them to set the stack level they want to see in their tests.

ElePT reviewed (commented):
LGTM! I just found a couple of deprecation warning assertions that I think would be good to clarify (maybe adding a comment explaining how they relate to this deprecation like you did in other cases).

mtreinish committed:
Merge remote-tracking branch 'origin/main' into deprecate-condition_cif

mtreinish committed:
Use private attribute in py->rust conversion

mtreinish committed:
Avoid deprecation warning in non-deprecated code

This commit fixes some places in the code where we were using the
deprecated functionality where it needed a path to persist the behavior
or where the deprecation warning became a performance bottleneck. The
code is updated accordingly and to catch issues like this in the future
the warning filters are adjusted to be more selective.

mtreinish committed:
Add missing test updates

ElePT committed:
Merge branch 'main' into deprecate-condition_cif

mtreinish committed:
Merge remote-tracking branch 'origin/main' into deprecate-condition_cif

mtreinish committed:
Add filter for aer's condition usage and dag drawer

mtreinish committed:
Fix lint

ElePT reviewed (approved):
LGTM

wshanks commented on releasenotes/notes/deprecate-condition_c_if-9548e5522814fe9c.yaml:
```diff
@@ -0,0 +1,38 @@
+
+        from qiskit.circuit import QuantumCircuit
+
+        qc = QuantumCircuit(2, 2)
+        qc.h(0)
+        qc.x(0).c_if(0, 1)
+        qc.z(1.c_if(1, 0)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
+
+    can be rewritten as::
+
+        qc = QuantumCircuit(2, 2)
+        qc.h(0)
+        with expected.if_test((expected.clbits[0], True)):
+            qc.x(0)
+        with expected.if_test((expected.clbits[1], False)):
+            qc.z(1)
+        qc.measure(0, 0)
+        qc.measure(1, 1)
```
I just started addressing this deprecation warning and noticed some issues here. I think it should be `qc.z(1)` (missing parenthesis). Also, I think `expected` should be `qc`?
